src/spi_bus.cpp

```diff
@@ -0,0 +1,364 @@
+/**
+ * @file motor_board.cpp
+ * @author Felix Widmaier (felix.widmaier@tuebingen.mpg.de)
+ * @author Manuel Wuthrich (manuel.wuthrich@gmail.com)
+ * @author Maximilien Naveau (maximilien.naveau@gmail.com)
+ * @brief This file implements the classes from
+ * "blmc_drivers/devices/motor_board.hpp"
+ * @version 0.1
+ * @date 2018-11-26
+ *
+ * @copyright Copyright (c) 2018
+ *
+ */
+
+#include "real_time_tools/timer.hpp"
+#include "blmc_drivers/devices/motor.hpp"
+#include "blmc_drivers/devices/spi_bus.hpp"
+
+
+namespace rt = real_time_tools;
+
+namespace blmc_drivers
+{
+
+SpiBus::SpiBus(
+    std::shared_ptr<MasterBoardInterface> master_board_interface,
+    const size_t& nb_udrivers, const size_t& history_length): DeviceInterface()
+{
+  // Save the ser input
+  main_board_interface_ = master_board_interface;
+  nb_udrivers_ = nb_udrivers;
+  history_length_ = history_length;
+  
+  // get the udriver from its index:
+  assert(N_SLAVES >= nb_udrivers_ && "The number of controlled udriver is not valid.");
+
+  // initialize the buffers
+  measurement_ = create_vector_of_pointers<MotorInterface::ScalarTimeseries>(
+      nb_udrivers_ * MotorBoardInterface::MeasurementIndex::measurement_count,
+      history_length_);
+
+  status_ = create_vector_of_pointers<MotorBoardInterface::StatusTimeseries>(
+      nb_udrivers_,
+      history_length_);
+
+  control_ = create_vector_of_pointers<MotorInterface::ScalarTimeseries>(
+      nb_udrivers_ * MotorBoardInterface::ControlIndex::control_count,
+      history_length);
+
+  command_ = create_vector_of_pointers<MotorBoardInterface::CommandTimeseries>(
+      nb_udrivers_, history_length);
+
+  sent_control_ = create_vector_of_pointers<MotorInterface::ScalarTimeseries>(
+      nb_udrivers_ * MotorBoardInterface::ControlIndex::control_count,
+      history_length);
+
+  sent_command_ = create_vector_of_pointers<MotorBoardInterface::CommandTimeseries>(
+      nb_udrivers_, history_length);
+
+  // start the communication loop
+  motors_are_paused_ = true;
+  is_loop_active_ = true;
+  rt_thread_.create_realtime_thread(&SpiBus::loop, this);
+}
+
+SpiBus::~SpiBus()
+{
+    is_loop_active_ = false;
+    rt_thread_.join();
+    // de-activate the robot
+    for(size_t i = 0 ; i < nb_udrivers_ ; ++i)
+    {
+        // Some aliases
+        MotorDriver& udriver = main_board_interface_->motor_drivers[i];
+        udriver.motor1->SetCurrentReference(0);
+        udriver.motor2->SetCurrentReference(0);
+        udriver.motor1->Disable();
+        udriver.motor2->Disable();
+        udriver.DisablePositionRolloverError();
+        udriver.Disable();
+    }
+}
+
+/**
+ * Output and status
+ */
+
+std::shared_ptr<const MotorBoardInterface::ScalarTimeseries>
+SpiBus::get_measurement(
+    const size_t udriver_id,
+    const MotorBoardInterface::MeasurementIndex& index) const
+{
+    return measurement_[udriver_id * 2 + index];
```

Comment:
i think this is wrong, i think this should probably be 
udriver_id * measurement_count + index.

in fact since you have to do this indexing in multiple places, i would probably add a function like
get_global_measurement_index(udriver_id, measurement_index)